src/Npgsql/TypeHandlers/ArrayHandler.cs

```diff
@@ -91,21 +91,59 @@ public ArrayHandler([CanBeNull] NpgsqlTypeHandler elementHandler, int lowerBound
         {
             // TODO: Throw SafeReadExceptions
             var t = typeof(TArray);
-            if (!t.IsArray)
-                throw new InvalidCastException($"Can't cast database type {PgDisplayName} to {typeof(TArray).Name}");
 
             // Getting an array
+            if (t.IsArray)
+            {
+                var elementType = t.GetElementType();
+                // We need to treat this as an actual array type, these need special treatment because of
+                // typing/generics reasons (there is no way to express "array of X" with generics
+                if (elementType == GetElementFieldType())
+                    return (TArray)await ReadAsObject(buf, len, async, fieldDescription);
+                if (elementType == GetElementPsvType())
+                    return (TArray)await ReadPsvAsObject(buf, len, async, fieldDescription);
+            }
+
+            if (t == typeof(List<TElement>))
+                return (TArray)await ReadAsList<TElement>(buf, async);
+
+            if (t.IsGenericType
+                && t.GetGenericTypeDefinition() == typeof(List<>)
+                && t.GenericTypeArguments[0] == GetElementPsvType())
+            {
+                return (TArray)await ReadPsvAsList(buf, async);
+            }
+                throw new InvalidCastException($"Can't cast database type {PgDisplayName} to {typeof(TArray).Name}");
+        }
+
+        protected async ValueTask<object> ReadAsList<TElement2>(NpgsqlReadBuffer buf, bool async)
+        {
+            await buf.Ensure(12, async);
+            var dimensions = buf.ReadInt32();
+
+            if (dimensions == 0)
+                return new List<TElement2>();
+            if (dimensions > 1)
+                throw new NotSupportedException($"Can't read multidimensional array as List<{typeof(TElement2).Name}>");
+
+            buf.ReadInt32();        // Has nulls. Not populated by PG?
+            var elementOID = buf.ReadUInt32();
+            // The following should hold but fails in test CopyTests.ReadBitString
+            //Debug.Assert(elementOID == ElementHandler.BackendType.OID);
```

Comment:
There are two bit types in PostgreSQL: `bit` and `varbit`. In the `ReadBitString` test you get the second one `oid` which `oid` is 1562. The element handler should provide multiple `oid` in such cases, but it's a different issue.

Reply:
/cc @roji

Reply:
This was a copy-pasted comment from ` Read\<TElement2>`.
I've removed it here.